src/axom/primal/geometry/BezierCurve.hpp

```diff
@@ -0,0 +1,335 @@
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file BezierCurve.hpp
+ *
+ * \brief A BezierCurve primitive
+ */
+
+#ifndef PRIMAL_BEZIERCURVE_HPP_
+#define PRIMAL_BEZIERCURVE_HPP_
+
+#include "axom/core.hpp"
+#include "axom/slic.hpp"
+
+#include "axom/primal/geometry/NumericArray.hpp"
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/geometry/Vector.hpp"
+#include "axom/primal/geometry/Segment.hpp"
+#include "axom/primal/geometry/BoundingBox.hpp"
+#include "axom/primal/geometry/OrientedBoundingBox.hpp"
+
+#include "axom/primal/operators/squared_distance.hpp"
+
+#include "fmt/fmt.hpp"
+#include <vector>
+#include <ostream>
+
+namespace axom
+{
+namespace primal
+{
+
+// Forward declare the templated classes and operator functions
+template < typename T, int NDIMS >
+class BezierCurve;
+
+/*! \brief Overloaded output operator for Bezier Curves*/
+template < typename T,int NDIMS >
+std::ostream& operator<<(std::ostream & os,
+                         const BezierCurve< T,NDIMS > & bCurve);
+
+/*!
+ * \class BezierCurve
+ *
+ * \brief Represents a Bezier curve defined by an array of control points
+ * \tparam T the coordinate type, e.g., double, float, etc.
+ * \tparam NDIMS the number of dimensions
+ *
+ * The order of a Bezier curve with N+1 control points is N.
+ * The curve is approximated by the control points,
+ * parametrized from t=0 to t=1.
+ */
+
+template < typename T,int NDIMS >
+class BezierCurve
+{
+public:
+  using PointType = Point< T,NDIMS >;
+  using VectorType = Vector< T,NDIMS >;
+  using NumArrayType = NumericArray< T,NDIMS >;
+  using SegmentType = Segment< T, NDIMS >;
+  using CoordsVec = std::vector< PointType >;
+  using BoundingBoxType = BoundingBox< T, NDIMS >;
+  using OrientedBoundingBoxType = OrientedBoundingBox< T, NDIMS >;
+
+public:
+
+  /*!
+   * \brief Constructor for a Bezier Curve that reserves space for
+   *  the given order of the curve
+   *
+   * \param [in] order the order of the resulting Bezier curve
+   * \pre order is greater than or equal to -1.
+   */
+  explicit BezierCurve(int ord = -1)
+  {
+    SLIC_ASSERT(ord >= -1);
+    const int sz = utilities::max(-1, ord+1);
+    m_controlPoints.resize(sz);
+  }
+
+  /*!
+   * \brief Constructor for an order \a ord= n Bezier curve
+   * from a list of coordinates:
+   * \verbatim {x_0, x_1, x_2,...,x_n,
+   *            y_0, y_1, y_2,...,y_n,
+   *            z_0, z_1, z_2,...,z_n}
+   *
+   * \param [in] pts an array with (n+1)*NDIMS entries, ordered by coordinate
+   * then by polynomial order
+   * \param [in] ord Polynomial order of the curve
+   * \pre order is greater than or equal to zero
+   */
+  BezierCurve(T* pts, int ord)
+  {
+    SLIC_ASSERT(pts != nullptr);
+    SLIC_ASSERT(ord >= 0);
+
+    const int sz = utilities::max(0, ord+1);
+    m_controlPoints.resize(sz);
+
+    for ( int p = 0 ; p <= ord ; p++)
+    {
+      auto& pt = m_controlPoints[p];
```

Comment:
This should just be `T` -- there is no strong need to use `auto` here.

Reply:
`auto` is the recommended return type for modern `C++`.  

E.g. see Scott Meyers' "Effective Modern C++" --> "Item 5: Prefer auto to explicit type declarations"

Reply:
Also, the return type would be ``PointType&`` since it is returning a point, rather than `T&`, a coordinate of a point.

Reply:
Then, this should definitely be `PointType&` to avoid confusion. I believe our coding guidelines discourage the use of `auto` unless it makes code cleaner, e.g., when using iterators etc.